### PR TITLE
feat(sdk): bucket multipart upload  

### PR DESCRIPTION
**Pull request by eladcon**
## Checklist

- [ ] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [ ] Description explains motivation and solution
- [ ] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingsdk/src/cloud/bucket.ts [hunk 1]
```diff
@@ -457,6 +469,21 @@ export interface IBucketClient {
...
    */
   rename(srcKey: string, dstKey: string): Promise<void>;
+
+  /**
+   * Initiate a multipart upload to a given key in the bucket.
+   * @param key The key to upload to
+   * @returns The upload id for the multipart upload
+   * @inflight
+   */
+  multipartUpload(key: string): Promise<string>;
```
Comment (eladb) [inline libs/wingsdk/src/cloud/bucket.ts:479]: Is the term "multipart" really helping here? I think the common developer won't really care that this is the underlying mechanism.

The docstrings should indicate that this will begin a multipart upload, but not needed in the name I think.

The return type can also be used to encode more meaning after you've landed on the right method. 

I would also return a struct (with the upload id inside) instead of a string so it's easier to correlate the start/stop.

So maybe something like this:

```js
struct MultipartUpload {
  id: str;
}

startUpload(key: str): MultipartUpload;

completeUpload(upload: MultipartUpload): void
```

Comment (eladb) [inline libs/wingsdk/src/cloud/bucket.ts:479]: Where do I specify the content?

File: libs/wingsdk/src/shared-aws/bucket.inflight.ts [hunk 2]
```diff
@@ -410,6 +413,21 @@ export class BucketClient implements IAwsBucketClient {
         });
         break;
       case BucketSignedUrlAction.UPLOAD:
+        if (opts?.uploadId !== undefined) {
```
Comment (eladb) [inline libs/wingsdk/src/shared-aws/bucket.inflight.ts:416]: How is this related to signed urls?

File: libs/wingsdk/src/shared-aws/permissions.ts [hunk 3]
```diff
@@ -160,6 +162,12 @@ export function calculateBucketPermissions(
   ) {
     actions.push("s3:CopyObject");
   }
+
+  // multipart upload signed url
+  if (ops.includes(cloud.BucketInflightMethods.SIGNED_URL)) {
```
Comment (eladb) [inline libs/wingsdk/src/shared-aws/permissions.ts:167]: ??

**Review by eladb — CHANGES_REQUESTED** [review]

**Comment by eladb** [comment]
Removed the word "support" from the title

File: libs/wingsdk/src/cloud/bucket.ts (hunk 1 above)
Comment (yoav-steinberg) [reply libs/wingsdk/src/cloud/bucket.ts]: this just initiates an upload, you the either use a signed url and an http put or the `uploadPart` method to specify the content.

File: libs/wingsdk/src/shared-aws/bucket.inflight.ts (hunk 2 above)
Comment (yoav-steinberg) [reply libs/wingsdk/src/shared-aws/bucket.inflight.ts]: You may generate a signed url specifically to upload a part of a multipart upload. When there's a multipart upload object attached to the sign request options then we know you want a signed url for a multipart upload.
